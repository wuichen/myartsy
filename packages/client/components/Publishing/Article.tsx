import { Theme } from "@artsy/palette"
import React from "react"

import { BannerWrapper } from "Components/Publishing/Banner/Banner"
import { PixelTracker } from "Components/Publishing/Display/ExternalTrackers"
import { EditorialFeature } from "Components/Publishing/EditorialFeature/EditorialFeature"
import ArticleWithFullScreen from "Components/Publishing/Layouts/ArticleWithFullScreen"
import { ClassicLayout } from "Components/Publishing/Layouts/ClassicLayout"
import { FeatureLayout } from "Components/Publishing/Layouts/FeatureLayout"
import { NewsLayout } from "Components/Publishing/Layouts/NewsLayout"
import { SeriesLayout } from "Components/Publishing/Layouts/SeriesLayout"
import { StandardLayout } from "Components/Publishing/Layouts/StandardLayout"
import { VideoLayout } from "Components/Publishing/Layouts/VideoLayout"
import { FullScreenProvider } from "Components/Publishing/Sections/FullscreenViewer/FullScreenProvider"
import { TooltipsDataProvider } from "Components/Publishing/ToolTip/TooltipsDataProvider"
import { ArticleData } from "Components/Publishing/Typings"
import { Bling as GPT } from "react-gpt"
import track, { TrackingProp } from "react-tracking"
import { MediaContextProvider } from "Utils/Responsive"
import Events from "../../Utils/Events"

GPT.enableSingleRequest()

export interface ArticleProps {
  articleSerial?: number
  article: ArticleData
  backgroundColor?: string
  color?: string
  customEditorial?: string
  relatedArticles?: any
  relatedArticlesForPanel?: any
  relatedArticlesForCanvas?: any
  renderTime?: number
  seriesArticle?: ArticleData
  isHovered?: boolean
  isLoggedIn?: boolean
  isMobile?: boolean
  isTablet?: boolean
  infiniteScrollEntrySlug?: string
  isSuper?: boolean
  isTruncated?: boolean
  emailSignupUrl?: string
  headerHeight?: string
  marginTop?: string | null
  showTooltips?: boolean
  showCollectionsRail?: boolean
  slideIndex?: number
  tracking?: TrackingProp
  closeViewer?: () => void
  viewerIsOpen?: boolean
  onOpenAuthModal?: (type: "register" | "login", config: object) => void
  onExpand?: () => void
  shouldAdRender?: boolean
}

@track(
  (props: ArticleProps) => {
    return {
      page: "Article",
      entity_type: "article",
      entity_id: props.article.id,
    }
  },
  {
    dispatch: data => Events.postEvent(data),
  }
)
export class Article extends React.Component<ArticleProps> {
  getArticleLayout = () => {
    const { article, customEditorial } = this.props

    if (customEditorial) {
      if (article.layout !== "series") {
        return (
          <FullScreenProvider>
            <ArticleWithFullScreen {...this.props}>
              <EditorialFeature {...this.props} />
            </ArticleWithFullScreen>
          </FullScreenProvider>
        )
      } else {
        return <EditorialFeature {...this.props} />
      }
    } else {
      switch (article.layout) {
        case "classic": {
          return <ClassicLayout {...this.props} />
        }
        case "feature": {
          return (
            <FullScreenProvider>
              <ArticleWithFullScreen {...this.props}>
                <FeatureLayout {...this.props} />
              </ArticleWithFullScreen>
            </FullScreenProvider>
          )
        }
        case "series": {
          return <SeriesLayout {...this.props} />
        }
        case "video": {
          return <VideoLayout {...this.props} />
        }
        case "news": {
          return <NewsLayout {...this.props} />
        }
        default: {
          return (
            <FullScreenProvider>
              <ArticleWithFullScreen {...this.props}>
                <StandardLayout {...this.props} />
              </ArticleWithFullScreen>
            </FullScreenProvider>
          )
        }
      }
    }
  }

  shouldRenderSignUpCta = () => {
    const { article, isLoggedIn, isTruncated, isMobile } = this.props

    return (
      isMobile && article.layout !== "series" && !isLoggedIn && !isTruncated
    )
  }

  sponsorPixelTrackingCode = article => {
    if (article.sponsor && article.sponsor.pixel_tracking_code) {
      return article.sponsor
    } else if (
      article.seriesArticle &&
      article.seriesArticle.sponsor &&
      article.seriesArticle.sponsor.pixel_tracking_code
    ) {
      return article.seriesArticle.sponsor
    }
  }

  render() {
    const { article } = this.props
    const trackingCode = this.sponsorPixelTrackingCode(article)

    return (
      <MediaContextProvider>
        <Theme>
          <TooltipsDataProvider {...this.props}>
            {this.getArticleLayout()}
            {trackingCode && (
              <PixelTracker unit={trackingCode} date={this.props.renderTime} />
            )}
            {this.shouldRenderSignUpCta() && (
              <BannerWrapper article={article} />
            )}
          </TooltipsDataProvider>
        </Theme>
      </MediaContextProvider>
    )
  }
}
