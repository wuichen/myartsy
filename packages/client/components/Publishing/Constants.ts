import { compact, last, uniq } from "lodash"
import { DateTime } from "luxon"
import url from "url"
import { ArticleData, DateFormat } from "../Publishing/Typings"

const APP_URL = process.env.APP_URL || "https://www.artsy.net"

/**
 * Matches for Email / Instant Articles
 */
export const EMAIL_REGEX = /^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

/**
 * The quality to request from image CDN
 */
export const GLOBAL_IMAGE_QUALITY = 80

/**
 * TODO: Eventually remove sizeMe
 */
export const SIZE_ME_REFRESH_RATE = 500

/**
 * Relative path to article
 */
export const getArticleHref = slug => `/article/${slug}`

/**
 * Absolute path to article
 */
export const getArticleFullHref = slug => `${APP_URL}/article/${slug}`

/**
 * Get the pre-slug part of URL
 */

export const getPreSlugPath = layout => {
  return ["standard", "feature"].includes(layout) ? "article" : layout
}

/**
 * Relative path to editorial entity
 */

export const getEditorialHref = (layout, slug) => {
  const layoutType = getPreSlugPath(layout)
  return `/${layoutType}/${slug}`
}

/**
 * Absolute path to editorial entity
 */
export const getFullEditorialHref = (layout, slug) => {
  const layoutType = getPreSlugPath(layout)
  return `${APP_URL}/${layoutType}/${slug}`
}

/**
 * Absolute path to artsy entity
 */
export const getFullArtsyHref = slug => `${APP_URL}/${slug}`

/**
 * ByLine helpers
 * TODO: Move this into some kind of utils folder
 */
export const getAuthorByline = (authors, isEditoral = true) => {
  const authorCount = Number(authors && authors.length)

  if (authorCount === 1) {
    return authors[0].name || ""
  } else if (authorCount > 1) {
    const names = authors.reduce((prev, curr, i) => {
      let delim
      const len = authors.length
      if (i === len - 1) {
        delim = " and "
      } else if (i === 0) {
        delim = ""
      } else {
        delim = ", "
      }
      return prev + delim + curr.name
    }, "")
    return names

    // No Author
  } else if (isEditoral) {
    return "Artsy Editors"
  }
}

export const getDate = (date, format: DateFormat = "default") => {
  const today = DateTime.local()
  const dateTime = DateTime.fromISO(date).setZone("America/New_York")
  const isToday = today.hasSame(dateTime, "day")
  const isThisYear = today.hasSame(dateTime, "year")
  const amPm = dateTime.hour >= 12 ? "pm" : "am"
  const minutes = dateTime.minute < 10 ? "0" + dateTime.minute : dateTime.minute
  const monthDay = `${dateTime.monthShort} ${dateTime.day}`
  const monthDayYear = `${dateTime.monthShort} ${dateTime.day}, ${
    dateTime.year
  }`
  let hour
  if (dateTime.hour > 12) {
    hour = dateTime.hour - 12
  } else if (dateTime.hour === 0) {
    hour = 12
  } else {
    hour = dateTime.hour
  }
  const time = `${hour}:${minutes}${amPm}`

  switch (format) {
    case "monthDay":
      return monthDay
    case "monthYear":
      return `${dateTime.monthShort} ${dateTime.year}`
    case "condensed":
      return monthDayYear
    case "verbose":
      const day = isToday ? "Today" : monthDayYear
      return `${day} at ${time}`
    case "news":
      return isToday ? "Today" : isThisYear ? monthDay : monthDayYear
    default:
      return `${dateTime.monthShort} ${dateTime.day}, ${dateTime.year} ${time}`
  }
}

export const getCurrentUnixTimestamp = () => DateTime.local().toMillis()

export const getMediaDate = article => {
  const { published_at, scheduled_publish_at, media } = article
  const { release_date } = media

  if (release_date) {
    return release_date
  } else {
    return published_at || scheduled_publish_at
  }
}

export const formatTime = time => {
  let minutes = Math.floor(time / 60) % 60
  let seconds = Math.floor(time % 60)
  minutes = minutes <= 0 ? 0 : minutes
  seconds = seconds <= 0 ? 0 : seconds

  const minutesStr = minutes < 10 ? "0" + minutes : minutes
  const secondsStr = seconds < 10 ? "0" + seconds : seconds
  return minutesStr + ":" + secondsStr
}

interface SlugsFromArticle {
  artists: string[]
  genes: string[]
}

export const getArtsySlugsFromArticle = (
  article: ArticleData
): SlugsFromArticle => {
  const articleBody = article.sections
    ? article.sections
        .map(section => {
          if (section.type === "text") {
            return section.body
          }
        })
        .join()
    : ""

  const artists = uniq(getArtsySlugsFromHTML(articleBody, "artist"))
  const genes = uniq(getArtsySlugsFromHTML(articleBody, "gene"))

  return {
    artists,
    genes,
  }
}

export const getArtsySlugsFromHTML = (
  html: string,
  model: string
): string[] => {
  const parser = new DOMParser()
  const doc = parser.parseFromString(html, "text/html")

  const slugs: string[] = []
  doc.querySelectorAll("a").forEach(anchor => {
    const href = anchor.getAttribute("href")
    if (href && href.match(`artsy.net/${model}`)) {
      slugs.push(last(url.parse(href).pathname.split("/")))
    }
  })

  return compact(slugs)
}
