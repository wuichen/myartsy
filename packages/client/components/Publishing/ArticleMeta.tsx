import { getFullEditorialHref } from "Components/Publishing/Constants"
import { ArticleData } from "Components/Publishing/Typings"
import React from "react"
import { data as sd } from "sharify"
import { crop } from "Utils/resizer"
import { createMediaStyle } from "Utils/Responsive"

const responsiveCss = createMediaStyle()

export const ArticleMeta: React.SFC<{
  article: ArticleData
}> = props => {
  const { article } = props
  const title = article.search_title || article.thumbnail_title || ""
  const titleExtension = article.layout === "news" ? "Artsy News" : "Artsy"
  const url = getFullEditorialHref(article.layout, article.slug)
  const socialTitle = article.social_title || article.thumbnail_title || ""
  const searchDescription = article.search_description || article.description
  const socialDescription = article.social_description || article.description
  const socialImage = article.social_image || article.thumbnail_image
  const authors = article.authors || article.contributing_authors || []
  const keywords = (article.keywords && article.keywords.join(", ")) || ""
  const sailthruKeywords = [`article${keywords && `, ${keywords}`}`]
  const emailMetadata = article.published && article.email_metadata

  if (article.featured) {
    sailthruKeywords.push("magazine")
  }

  return (
    <>
      <title>{`${title} - ${titleExtension}`}</title>
      <link rel="canonical" href={url} />
      <meta name="description" content={searchDescription} />

      {/** amp url, do we need? */
      article.featured &&
        article.published &&
        ["standard", "feature"].includes(article.layout) && (
          <link rel="amphtml" href={`${url}/amp`} />
        )}

      {/** OpenGraph / Facebook */}
      <meta property="og:site_name" content="Artsy" />
      <meta property="og:title" name="og:title" content={socialTitle} />
      <meta property="og:url" content={url} />
      <meta property="og:type" content="article" />
      <meta property="og:published_time" content={article.published_at} />
      <meta property="og:description" content={socialDescription} />
      <meta property="og:image" content={socialImage} />
      <meta property="fb:app_id" content={sd.FACEBOOK_ID} />
      <meta property="article:published_time" content={article.published_at} />
      <meta
        property="article:publisher"
        content="https://www.facebook.com/artsy/"
      />

      {/** Twitter */}
      <meta name="twitter:site" content="@artsy" />
      <meta property="twitter:card" content="summary_large_image" />
      <meta property="twitter:url" content={url} />
      <meta property="twitter:description" content={socialDescription} />
      <meta property="twitter:image" content={socialImage} />
      <meta name="twitter:title" content={socialTitle} />

      {/** Sailthru */
      emailMetadata && (
        <>
          <meta name="sailthru.date" content={article.published_at} />
          {emailMetadata.headline && (
            <meta name="sailthru.title" content={emailMetadata.headline} />
          )}
          {emailMetadata.author && (
            <meta name="sailthru.author" content={emailMetadata.author} />
          )}
          {emailMetadata.image_url && (
            <>
              <meta
                name="sailthru.image.full"
                content={crop(emailMetadata.image_url, {
                  width: 1200,
                  height: 800,
                })}
              />
              <meta
                name="sailthru.image.thumb"
                content={crop(emailMetadata.image_url, {
                  width: 600,
                  height: 400,
                })}
              />
            </>
          )}
          <meta name="sailthru.tags" content={sailthruKeywords.join(", ")} />
        </>
      )}

      {keywords.length && (
        <>
          <meta property="news_keywords" content={keywords} />
          <meta name="keywords" content={keywords} />
          <meta property="article:tag" content={keywords} />
        </>
      )}

      {authors.map(
        (author, i) =>
          author.name && <meta key={i} name="author" content={author.name} />
      )}
      {!article.indexable && <meta name="robots" content="noindex" />}
      <meta name="viewport" content="width=device-width, initial-scale=1.0" />
      <style>{responsiveCss}</style>
    </>
  )
}
