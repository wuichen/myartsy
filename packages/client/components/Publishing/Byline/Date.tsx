import { Sans, SansSize } from "@artsy/palette"
import React from "react"
import styled from "styled-components"
import { getDate } from "../Constants"
import { BylineLayout, DateFormat } from "../Typings"

interface DateProps {
  date?: string
  format?: DateFormat
  layout?: BylineLayout
  size?: SansSize
}

export const Date: React.SFC<DateProps> = props => {
  const { date, format, layout, size } = props
  const condensed = layout === "condensed"
  const dateFormat = format || (condensed ? "condensed" : "default")
  const fontSize = size ? size : condensed ? "2" : "3t"

  return (
    <DateContainer condensed={condensed}>
      <Sans size={fontSize} weight="medium">
        {getDate(date, dateFormat)}
      </Sans>
    </DateContainer>
  )
}

export const DateContainer = styled.div.attrs<{ condensed?: boolean }>({})`
  margin: 5px 20px 0 0;
  white-space: nowrap;
`
