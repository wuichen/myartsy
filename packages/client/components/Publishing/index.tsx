// Components
export { Article } from "./Article"
export { Artwork } from "./Sections/Artwork"
export { Authors } from "./Sections/Authors"
export { Caption } from "./Sections/Caption"
export { Embed } from "./Sections/Embed"
export { FullscreenViewer } from "./Sections/FullscreenViewer/FullscreenViewer"
export { Header } from "./Header/Header"
export { InstantArticleEmailSignup } from "./Email/InstantArticleEmailSignup"
export { Image } from "./Sections/Image"
export { ImageCollection } from "./Sections/ImageCollection"
export { ImageSetPreview } from "./Sections/ImageSetPreview"
export {
  ImageSetPreviewClassic,
} from "./Sections/ImageSetPreview/ImageSetPreviewClassic"
export { Nav } from "./Nav/Nav"
export { SeriesLayout } from "./Layouts/SeriesLayout"
export { Text } from "./Sections/Text"
export { Video } from "./Sections/Video"
export { VideoLayout } from "./Layouts/VideoLayout"
export { VideoCover } from "./Video/VideoCover"
export { VideoAbout } from "./Video/VideoAbout"
export { PartnerInline } from "./Partner/PartnerInline"
export { PartnerBlock } from "./Partner/PartnerBlock"
export {
  RelatedArticlesCanvas,
} from "./RelatedArticles/Canvas/RelatedArticlesCanvas"
export {
  RelatedArticlesPanel,
} from "./RelatedArticles/Panel/RelatedArticlesPanel"

// Icon SVGs
export { IconArtist } from "./Icon/IconArtist"
export { IconBlockquote } from "./Icon/IconBlockquote"
export { IconClearFormatting } from "./Icon/IconClearFormatting"
export { IconDrag } from "./Icon/IconDrag"
export { IconEditEmbed } from "./Icon/IconEditEmbed"
export { IconEditImages } from "./Icon/IconEditImages"
export { IconEditSection } from "./Icon/IconEditSection"
export { IconEditText } from "./Icon/IconEditText"
export { IconEditVideo } from "./Icon/IconEditVideo"
export { IconHamburger } from "./Icon/IconHamburger"
export { IconHeroImage } from "./Icon/IconHeroImage"
export { IconHeroVideo } from "./Icon/IconHeroVideo"
export { IconImageFullscreen } from "./Icon/IconImageFullscreen"
export { IconImageSet } from "./Icon/IconImageSet"
export { IconLayoutFullscreen } from "./Icon/IconLayoutFullscreen"
export { IconLayoutSplit } from "./Icon/IconLayoutSplit"
export { IconLayoutText } from "./Icon/IconLayoutText"
export { IconLayoutBasic } from "./Icon/IconLayoutBasic"
export { IconLink } from "./Icon/IconLink"
export { IconOrderedList } from "./Icon/IconOrderedList"
export { IconVideoPlay } from "./Icon/IconVideoPlay"
export { IconRemove } from "./Icon/IconRemove"
export { IconUnorderedList } from "./Icon/IconUnorderedList"
export { IconLock } from "./Icon/IconLock"

// FIXME: Refactor out SizeMe; see https://github.com/ctrlplusb/react-sizeme#server-side-rendering
import sizeMe from "react-sizeme"
sizeMe.noPlaceholders = true

// Constants
import * as AllConstants from "./Constants"
export const Constants = AllConstants
