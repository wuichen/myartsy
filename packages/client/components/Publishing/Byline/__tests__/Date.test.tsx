import { getDate } from "Components/Publishing/Constants"
import { mount } from "enzyme"
import "jest-styled-components"
import { DateTime } from "luxon"
import React from "react"
import renderer from "react-test-renderer"
import { Date } from "../Date"

describe("Date", () => {
  const date = "2017-05-19T13:09:18.567Z"

  describe("Snapshots", () => {
    it("renders a date", () => {
      const snapshot = renderer.create(<Date date={date} layout="split" />)
      expect(snapshot).toMatchSnapshot()
    })

    it("renders a condensed date", () => {
      const snapshot = renderer.create(<Date date={date} layout="condensed" />)
      expect(snapshot).toMatchSnapshot()
    })
  })

  describe("Unit", () => {
    it("renders the date", () => {
      const component = mount(<Date date={date} layout="split" />)
      expect(component.text()).toBe("May 19, 2017 9:09am")
    })

    it("renders condensed date", () => {
      const component = mount(<Date date={date} layout="condensed" />)
      expect(component.text()).toBe("May 19, 2017")
    })
  })
})

describe("#getDate", () => {
  const timestamp = "2017-02-22T19:22:05.709Z"
  const expectedFormattedDates = {
    monthYear: "Feb 2017",
    condensed: "Feb 22, 2017",
    verbose: "Feb 22, 2017 at 2:22pm",
    verboseToday: "Today at 2:22pm",
    default: "Feb 22, 2017 2:22pm",
  }

  it("returns the correct date for monthYear format", () => {
    const date = getDate(timestamp, "monthYear")
    expect(date).toEqual(expectedFormattedDates.monthYear)
  })

  it("returns the correct date for condensed format", () => {
    const date = getDate(timestamp, "condensed")
    expect(date).toEqual(expectedFormattedDates.condensed)
  })

  it("returns the correct date for verbose format", () => {
    const date = getDate(timestamp, "verbose")
    expect(date).toEqual(expectedFormattedDates.verbose)
  })

  it("returns the correct date for default format", () => {
    const date = getDate(timestamp, "default")
    expect(date).toEqual(expectedFormattedDates.default)
  })

  // FIXME: Reenable test (fails CI at certain times)
  xit("returns the correct verbose date for same-day timestamps", () => {
    let today = DateTime.local().setZone("America/New_York")
    today = today.set({
      hour: 14,
      minute: 22,
    })

    const date = getDate(today, "verbose")
    expect(date).toEqual(expectedFormattedDates.verboseToday)
  })
})
