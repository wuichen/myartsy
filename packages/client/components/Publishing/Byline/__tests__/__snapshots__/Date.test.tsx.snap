// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Date Snapshots renders a condensed date 1`] = `
.c1 {
  font-family: Unica77LLWebMedium,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-size: 12px;
  line-height: 16px;
}

.c0 {
  margin: 5px 20px 0 0;
  white-space: nowrap;
}

<div
  className="c0"
>
  <div
    className="c1"
    fontFamily={
      Object {
        "fontFamily": "Unica77LLWebMedium, 'Helvetica Neue', Helvetica, Arial, sans-serif",
        "fontWeight": 500,
      }
    }
    fontSize="12px"
  >
    May 19, 2017
  </div>
</div>
`;

exports[`Date Snapshots renders a date 1`] = `
.c1 {
  font-family: Unica77LLWebMedium,'Helvetica Neue',Helvetica,Arial,sans-serif;
  font-weight: 500;
  font-size: 14px;
  line-height: 20px;
}

.c0 {
  margin: 5px 20px 0 0;
  white-space: nowrap;
}

<div
  className="c0"
>
  <div
    className="c1"
    fontFamily={
      Object {
        "fontFamily": "Unica77LLWebMedium, 'Helvetica Neue', Helvetica, Arial, sans-serif",
        "fontWeight": 500,
      }
    }
    fontSize="14px"
  >
    May 19, 2017 9:09am
  </div>
</div>
`;
